File: .claude/skills/aggy-import/scripts/validate_layout.py
#!/usr/bin/env python3
"""Aggy レイアウトJSON バリデーター

Usage:
    python validate_layout.py <layout.json> [--csv <rawdata.csv>]

layout.schema.json に基づいてレイアウトJSONを検証する。
--csv を指定すると、CSVヘッダーとの整合性もチェックする。
"""
import json
import csv
import sys
import os


def validate_layout(layout: list, schema_path: str | None = None) -> list[str]:
    """レイアウトJSONをスキーマルールに基づいて検証する。"""
    errors = []

    if not isinstance(layout, list):
        return ["ルート要素がJSON配列ではありません"]

    if len(layout) == 0:
        return ["レイアウトが空です"]

    seen_keys = set()

    for i, entry in enumerate(layout):
        prefix = f"[{i}]"

        if not isinstance(entry, dict):
            errors.append(f"{prefix}: オブジェクトではありません")
            continue

        # key: 必須、文字列
        key = entry.get("key")
        if key is None:
            errors.append(f"{prefix}: 'key' がありません")
        elif not isinstance(key, str) or not key:
            errors.append(f"{prefix}: 'key' が空または文字列ではありません")
        else:
            if key in seen_keys:
                errors.append(f"{prefix}: key '{key}' が重複しています")
            seen_keys.add(key)

        # type: 必須、SA/MA/WEIGHT/DATE
        typ = entry.get("type")
        valid_types = {"SA", "MA", "WEIGHT", "DATE"}
        if typ is None:
            errors.append(f"{prefix}: 'type' がありません")
            continue
        elif typ not in valid_types:
            errors.append(f"{prefix}: type '{typ}' は無効です（有効: {valid_types}）")
            continue

        # 許可されるフィールドのチェック
        allowed = {"key", "type", "label", "items", "granularity"}
        extra = set(entry.keys()) - allowed
        if extra:
            errors.append(f"{prefix}: 不明なフィールド: {extra}")

        # 型別ルール
        if typ in ("SA", "MA"):
            _validate_sa_ma(entry, prefix, typ, errors)
        elif typ == "WEIGHT":
            _validate_weight(entry, prefix, errors)
        elif typ == "DATE":
            _validate_date(entry, prefix, errors)

    return errors


def _validate_sa_ma(entry: dict, prefix: str, typ: str, errors: list):
    """SA/MA共通の検証。"""
    items = entry.get("items")
    if items is None:
        errors.append(f"{prefix}: {typ} には 'items' が必須です")
        return
    if not isinstance(items, list):
        errors.append(f"{prefix}: 'items' が配列ではありません")
        return
    if len(items) == 0:
        errors.append(f"{prefix}: 'items' が空です（1つ以上必要）")
        return

    seen_codes = set()
    for j, item in enumerate(items):
        ip = f"{prefix}.items[{j}]"
        if not isinstance(item, dict):
            errors.append(f"{ip}: オブジェクトではありません")
            continue

        # code: 必須、文字列
        code = item.get("code")
        if code is None:
            errors.append(f"{ip}: 'code' がありません")
        elif not isinstance(code, str):
            errors.append(f"{ip}: 'code' が文字列ではありません（値: {code!r}, 型: {type(code).__name__}）")
        else:
            if code in seen_codes:
                errors.append(f"{ip}: code '{code}' が重複しています")
            seen_codes.add(code)

        # label: 必須、文字列
        label = item.get("label")
        if label is None:
            errors.append(f"{ip}: 'label' がありません")
        elif not isinstance(label, str):
            errors.append(f"{ip}: 'label' が文字列ではありません")

        # 余分なフィールド
        extra = set(item.keys()) - {"code", "label"}
        if extra:
            errors.append(f"{ip}: 不明なフィールド: {extra}")


def _validate_weight(entry: dict, prefix: str, errors: list):
    """WEIGHT型の検証。"""
    if "items" in entry:
        errors.append(f"{prefix}: WEIGHT に 'items' は不要です")
    if "granularity" in entry:
        errors.append(f"{prefix}: WEIGHT に 'granularity' は不要です")


def _validate_date(entry: dict, prefix: str, errors: list):
    """DATE型の検証。"""
    if "items" in entry:
        errors.append(f"{prefix}: DATE に 'items' は不要です")

    gran = entry.get("granularity")
    valid_gran = {"year", "month", "week", "day"}
    if gran is None:
        errors.append(f"{prefix}: DATE には 'granularity' が必須です")
    elif gran not in valid_gran:
        errors.append(f"{prefix}: granularity '{gran}' は無効です（有効: {valid_gran}）")


def check_csv_consistency(layout: list, csv_path: str) -> list[str]:
    """CSVヘッダーとレイアウトJSONの整合性をチェックする。"""
    errors = []

    with open(csv_path, encoding="utf-8") as f:
        reader = csv.reader(f)
        headers = next(reader)

    header_set = set(headers)
    expected = set()

    for entry in layout:
        typ = entry.get("type")
        key = entry.get("key", "")

        if typ in ("SA", "WEIGHT", "DATE"):
            expected.add(key)
            if key not in header_set:
                errors.append(f"レイアウトの key '{key}' ({typ}) がCSVヘッダーにありません")

        elif typ == "MA":
            for item in entry.get("items", []):
                col = f"{key}_{item['code']}"
                expected.add(col)
                if col not in header_set:
                    errors.append(f"MA列 '{col}' がCSVヘッダーにありません")

    extra = header_set - expected
    if extra:
        errors.append(f"レイアウト未定義のCSV列があります: {sorted(extra)}")

    return errors


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    layout_path = sys.argv[1]
    csv_path = None

    if "--csv" in sys.argv:
        idx = sys.argv.index("--csv")
        if idx + 1 < len(sys.argv):
            csv_path = sys.argv[idx + 1]

    # レイアウトJSON読み込み
    try:
        with open(layout_path, encoding="utf-8") as f:
            layout = json.load(f)
    except json.JSONDecodeError as e:
        print(f"FAIL: JSONパースエラー: {e}")
        sys.exit(1)
    except FileNotFoundError:
        print(f"FAIL: ファイルが見つかりません: {layout_path}")
        sys.exit(1)

    # スキーマバリデーション
    errors = validate_layout(layout)

    # CSV整合性チェック
    csv_errors = []
    if csv_path:
        try:
            csv_errors = check_csv_consistency(layout, csv_path)
        except FileNotFoundError:
            csv_errors = [f"CSVファイルが見つかりません: {csv_path}"]

    all_errors = errors + csv_errors

    if all_errors:
        print(f"FAIL: {len(all_errors)} 件のエラー")
        for e in all_errors:
            print(f"  - {e}")
        sys.exit(1)
    else:
        # サマリー
        type_counts = {}
        for entry in layout:
            t = entry.get("type", "?")
            type_counts[t] = type_counts.get(t, 0) + 1
        summary = ", ".join(f"{t} x{c}" for t, c in sorted(type_counts.items()))
        print(f"OK: {len(layout)} エントリ ({summary})")
        if csv_path:
            print(f"OK: CSV整合性チェック通過")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: .claude/skills/aggy-import/scripts/validate_layout.mjs
#!/usr/bin/env node
/**
 * Aggy レイアウトJSON バリデーター
 *
 * Usage:
 *     node validate_layout.mjs <layout.json> [--csv <rawdata.csv>]
 *
 * layout.schema.json に基づいてレイアウトJSONを検証する。
 * --csv を指定すると、CSVヘッダーとの整合性もチェックする。
 */
import { readFileSync } from "fs";
import { resolve } from "path";

// --- Layout validation ---

function validateLayout(layout) {
  const errors = [];

  if (!Array.isArray(layout)) {
    return ["ルート要素がJSON配列ではありません"];
  }
  if (layout.length === 0) {
    return ["レイアウトが空です"];
  }

  const seenKeys = new Set();

  for (let i = 0; i < layout.length; i++) {
    const entry = layout[i];
    const prefix = `[${i}]`;

    if (typeof entry !== "object" || entry === null || Array.isArray(entry)) {
      errors.push(`${prefix}: オブジェクトではありません`);
      continue;
    }

    // key: 必須、文字列
    const key = entry.key;
    if (key === undefined || key === null) {
      errors.push(`${prefix}: 'key' がありません`);
    } else if (typeof key !== "string" || key === "") {
      errors.push(`${prefix}: 'key' が空または文字列ではありません`);
    } else {
      if (seenKeys.has(key)) {
        errors.push(`${prefix}: key '${key}' が重複しています`);
      }
      seenKeys.add(key);
    }

    // type: 必須、SA/MA/WEIGHT/DATE
    const typ = entry.type;
    const validTypes = new Set(["SA", "MA", "WEIGHT", "DATE"]);
    if (typ === undefined || typ === null) {
      errors.push(`${prefix}: 'type' がありません`);
      continue;
    }
    if (!validTypes.has(typ)) {
      errors.push(`${prefix}: type '${typ}' は無効です（有効: ${[...validTypes].join(", ")}）`);
      continue;
    }

    // 許可されるフィールドのチェック
    const allowed = new Set(["key", "type", "label", "items", "granularity"]);
    const extra = Object.keys(entry).filter((k) => !allowed.has(k));
    if (extra.length > 0) {
      errors.push(`${prefix}: 不明なフィールド: {${extra.join(", ")}}`);
    }

    // 型別ルール
    if (typ === "SA" || typ === "MA") {
      validateSaMa(entry, prefix, typ, errors);
    } else if (typ === "WEIGHT") {
      validateWeight(entry, prefix, errors);
    } else if (typ === "DATE") {
      validateDate(entry, prefix, errors);
    }
  }

  return errors;
}

function validateSaMa(entry, prefix, typ, errors) {
  const items = entry.items;
  if (items === undefined || items === null) {
    errors.push(`${prefix}: ${typ} には 'items' が必須です`);
    return;
  }
  if (!Array.isArray(items)) {
    errors.push(`${prefix}: 'items' が配列ではありません`);
    return;
  }
  if (items.length === 0) {
    errors.push(`${prefix}: 'items' が空です（1つ以上必要）`);
    return;
  }

  const seenCodes = new Set();
  for (let j = 0; j < items.length; j++) {
    const item = items[j];
    const ip = `${prefix}.items[${j}]`;

    if (typeof item !== "object" || item === null || Array.isArray(item)) {
      errors.push(`${ip}: オブジェクトではありません`);
      continue;
    }

    // code: 必須、文字列
    const code = item.code;
    if (code === undefined || code === null) {
      errors.push(`${ip}: 'code' がありません`);
    } else if (typeof code !== "string") {
      errors.push(
        `${ip}: 'code' が文字列ではありません（値: ${JSON.stringify(code)}, 型: ${typeof code}）`,
      );
    } else {
      if (seenCodes.has(code)) {
        errors.push(`${ip}: code '${code}' が重複しています`);
      }
      seenCodes.add(code);
    }

    // label: 必須、文字列
    const label = item.label;
    if (label === undefined || label === null) {
      errors.push(`${ip}: 'label' がありません`);
    } else if (typeof label !== "string") {
      errors.push(`${ip}: 'label' が文字列ではありません`);
    }

    // 余分なフィールド
    const extraFields = Object.keys(item).filter((k) => k !== "code" && k !== "label");
    if (extraFields.length > 0) {
      errors.push(`${ip}: 不明なフィールド: {${extraFields.join(", ")}}`);
    }
  }
}

function validateWeight(entry, prefix, errors) {
  if ("items" in entry) {
    errors.push(`${prefix}: WEIGHT に 'items' は不要です`);
  }
  if ("granularity" in entry) {
    errors.push(`${prefix}: WEIGHT に 'granularity' は不要です`);
  }
}

function validateDate(entry, prefix, errors) {
  if ("items" in entry) {
    errors.push(`${prefix}: DATE に 'items' は不要です`);
  }

  const gran = entry.granularity;
  const validGran = new Set(["year", "month", "week", "day"]);
  if (gran === undefined || gran === null) {
    errors.push(`${prefix}: DATE には 'granularity' が必須です`);
  } else if (!validGran.has(gran)) {
    errors.push(
      `${prefix}: granularity '${gran}' は無効です（有効: ${[...validGran].join(", ")}）`,
    );
  }
}

// --- CSV consistency check ---

function parseCsvFirstLine(csvPath) {
  const content = readFileSync(csvPath, "utf-8");
  // Handle both \r\n and \n line endings
  const firstLine = content.split(/\r?\n/)[0];
  // Simple CSV header parse (no quotes handling needed for headers typically)
  return firstLine.split(",").map((h) => h.trim().replace(/^"|"$/g, ""));
}

function checkCsvConsistency(layout, csvPath) {
  const errors = [];
  const headers = parseCsvFirstLine(csvPath);
  const headerSet = new Set(headers);
  const expected = new Set();

  for (const entry of layout) {
    const typ = entry.type;
    const key = entry.key || "";

    if (typ === "SA" || typ === "WEIGHT" || typ === "DATE") {
      expected.add(key);
      if (!headerSet.has(key)) {
        errors.push(`レイアウトの key '${key}' (${typ}) がCSVヘッダーにありません`);
      }
    } else if (typ === "MA") {
      for (const item of entry.items || []) {
        const col = `${key}_${item.code}`;
        expected.add(col);
        if (!headerSet.has(col)) {
          errors.push(`MA列 '${col}' がCSVヘッダーにありません`);
        }
      }
    }
  }

  const extra = [...headerSet].filter((h) => !expected.has(h)).sort();
  if (extra.length > 0) {
    errors.push(`レイアウト未定義のCSV列があります: [${extra.join(", ")}]`);
  }

  return errors;
}

// --- Main ---

function printUsage() {
  console.log(`Aggy レイアウトJSON バリデーター

Usage:
    node validate_layout.mjs <layout.json> [--csv <rawdata.csv>]

layout.schema.json に基づいてレイアウトJSONを検証する。
--csv を指定すると、CSVヘッダーとの整合性もチェックする。`);
}

function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes("--help") || args.includes("-h")) {
    printUsage();
    process.exit(args.length === 0 ? 1 : 0);
  }

  const layoutPath = resolve(args[0]);
  let csvPath = null;

  const csvIdx = args.indexOf("--csv");
  if (csvIdx !== -1 && csvIdx + 1 < args.length) {
    csvPath = resolve(args[csvIdx + 1]);
  }

  // レイアウトJSON読み込み
  let layout;
  try {
    const raw = readFileSync(layoutPath, "utf-8");
    layout = JSON.parse(raw);
  } catch (e) {
    if (e.code === "ENOENT") {
      console.log(`FAIL: ファイルが見つかりません: ${layoutPath}`);
    } else {
      console.log(`FAIL: JSONパースエラー: ${e.message}`);
    }
    process.exit(1);
  }

  // スキーマバリデーション
  const errors = validateLayout(layout);

  // CSV整合性チェック
  let csvErrors = [];
  if (csvPath) {
    try {
      csvErrors = checkCsvConsistency(layout, csvPath);
    } catch (e) {
      if (e.code === "ENOENT") {
        csvErrors = [`CSVファイルが見つかりません: ${csvPath}`];
      } else {
        csvErrors = [`CSV読み込みエラー: ${e.message}`];
      }
    }
  }

  const allErrors = [...errors, ...csvErrors];

  if (allErrors.length > 0) {
    console.log(`FAIL: ${allErrors.length} 件のエラー`);
    for (const e of allErrors) {
      console.log(`  - ${e}`);
    }
    process.exit(1);
  } else {
    // サマリー
    const typeCounts = {};
    for (const entry of layout) {
      const t = entry.type || "?";
      typeCounts[t] = (typeCounts[t] || 0) + 1;
    }
    const summary = Object.entries(typeCounts)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([t, c]) => `${t} x${c}`)
      .join(", ");
    console.log(`OK: ${layout.length} エントリ (${summary})`);
    if (csvPath) {
      console.log("OK: CSV整合性チェック通過");
    }
    process.exit(0);
  }
}

main();
